api/src/groups/api.py

```diff
@@ -5,527 +5,507 @@
+# ---- Manager Type
 
-    if group is None:
-        return jsonify(msg="Group not found"), 404
+manager_type_schema = ManagerTypeSchema()
 
-    setattr(group, 'active', False)
-    db.session.commit()
-    return jsonify(group_schema.dump(group))
+@groups.route('/manager-types', methods=['POST'])
+def create_manager_type():
+    try:
+        valid_manager_type = manager_type_schema.load(request.json)
+    except ValidationError as err:
+        return jsonify(err.messages), 422
 
+    new_manager_type = ManagerType(**valid_manager_type)
+    db.session.add(new_manager_type)
+    db.session.commit()
+    return jsonify(manager_type_schema.dump(new_manager_type)), 201
 
-# ---- Meeting
+@groups.route('/manager-types/<int:manager_type_id>', methods=['GET'])
+def read_one_manager_type(manager_type_id):
+    manager_type = db.session.query(ManagerType).filter_by(id=manager_type_id).first()
+    if not manager_type:
+        return jsonify(f"ManagerType with id #{manager_type_id} does not exist."), 404
 
-meeting_schema = MeetingSchema()
+    return jsonify(manager_type_schema.dump(manager_type))
 
+@groups.route('/manager-types', methods=['GET'])
+def read_all_manager_types():
+    query = db.session.query(ManagerType)
+    try:
+        manager_types = get_all_queried_entities(query, request.args)
+    except QueryArgumentError as e:
+        return jsonify(e.message), e.code
+    return jsonify(manager_type_schema.dump(manager_types, many=True))
 
-@groups.route('/meetings', methods=['POST'])
+@groups.route('/manager-types/<int:manager_type_id>', methods=['PATCH'])
 @jwt_required
-def create_meeting():
-    if 'active' not in request.json.keys():
-        request.json['active'] = True
+def update_manager_type(manager_type_id):
+    manager_type_schema = ManagerTypeSchema()
+
     try:
-        valid_meeting = meeting_schema.load(request.json)
+        valid_attributes = manager_type_schema.load(request.json, partial=True)
     except ValidationError as err:
         return jsonify(err.messages), 422
 
-    new_meeting = Meeting(**valid_meeting)
-    db.session.add(new_meeting)
-    db.session.commit()
-    return jsonify(meeting_schema.dump(new_meeting)), 201
-
+    return modify_entity(ManagerType, manager_type_schema, manager_type_id, valid_attributes)
 
-@groups.route('/meetings')
-def read_all_meetings():
-    result = db.session.query(Meeting).all()
 
-    if result == []:
-        return jsonify(msg="No meetings found"), 404
+@groups.route('/manager-types/<int:manager_type_id>', methods=['DELETE'])
+@jwt_required
+def delete_manager_type(manager_type_id):
+    manager_type = db.session.query(ManagerType).filter_by(id=manager_type_id).first()
 
-    for r in result:
-        r.address = r.address
-    return jsonify(meeting_schema.dump(result, many=True))
+    if not manager_type:
+        return jsonify(f"ManagerType with id #{manager_type_id} does not exist."), 404
 
+    db.session.delete(manager_type)
```

Comment:
In order to delete a manager type, we would have to make sure no `Manager` records were using that type. Should we:
* Only allow this if there are no managers using this type?
* Make the manager type FK be nullable in the `Manager` model class?
* Allow the user to choose a different manager type that will replace instances of the manager type that we're about to delete?

Reply:
Yeah, I think you made a good point... Currently it will probably fail with an internal server error (and the samething for a `GroupType`). 

For the three options, here are just some of my thoughts:
1. I can see this mechanism working, but it seems a little strict, in that when someone tries to modify `manager_types` they would have to remember to first make a new type and change all managers to that type _before_ deleting the existing type. But I guess we can do something in the UI to make it more user-friendly as well.
1. This is just an alternative to the previous option, which offers more flexibility.
1. This can perhaps be a logic on the UI side, where the user is prompted to enter a new name for the manager type to be deleted. On the API side, this can be done by making a PATCH request to the same endpoint.

So it seems to me that for API, the decision is just between the first and second, about whether we need some extra flexbility by making the FK nullable. What do you prefer?
I will implement the change once we reach a conclusion on the design.

Reply:
Good analysis. How about:
1. Server side: leave things alone; it's good to have the referential integrity that you get with a non-nullable FK.
1. UI side: if there are instances of `Manager` with the given `Manager Type`, implement UI that allows the user to change those instances to be of a different `Manager Type`.

The UI could, for example, list everyone who has the "doomed" type, along with a dropdown of other management types, allowing the user to choose a new management type for each such manager. A simpler alternative would be to only allow a "bulk" change, where _everyone_ with the doomed type would be updated to a different type.

Which UI approach do you think would be better for the end user?

We need a new issue for this change.